File: BattlrNet/ClientApp/Vector.ts
import IPoint from "./IPoint";

export default class Vector implements IPoint {
    public x: number;
    public y: number;

    constructor(x?: number, y?: number) {
        this.x = x || 0;
        this.y = y || 0;
    }

    /** Modifies this vector */
    public copyFrom(other: Vector) {
        this.x = other.x;
        this.y = other.y;
        return this;
    }

    /** Modifies this vector */
    public add(other: Vector) {
        this.x += other.x;
        this.y += other.y;
        return this;
    }

    /** Modifies this vector */
    public subtract(other: Vector) {
        this.x -= other.x;
        this.y -= other.y;
        return this;
    }

    /** Modifies this vector */
    public multiply(value: number) {
        this.x *= value;
        this.y *= value;
        return this;
    }

    /** Modifies this vector */
    public round() {
        this.x = Math.round(this.x);
        this.y = Math.round(this.y);
        return this;
    }

    public clone() {
        return new Vector(this.x, this.y);
    }

    public dot(other: Vector) {
        return this.x * other.x + this.y * other.y;
    }

    public length() {
        return Math.sqrt(this.x * this.x + this.y * this.y);
    }

    public lengthSquared() {
        return this.x * this.x + this.y * this.y;
    }

    public distance(other: Vector) {
        const x = this.x - other.x;
        const y = this.y - other.y;
        return Math.sqrt(x * x + y * y);
    }

    public distanceSquared(other: Vector) {
        const x = this.x - other.x;
        const y = this.y - other.y;
        return x * x + y * y;
    }

    static add(a: Vector, b: Vector, out?: Vector): Vector {
        if (!out) out = new Vector();
        out.x = a.x + b.x;
        out.y = a.y + b.y;
        return out;
    }

    static subtract(a: Vector, b: Vector, out?: Vector): Vector {
        if (!out) out = new Vector();
        out.x = a.x - b.x;
        out.y = a.y - b.y;
        return out;
    }

    static multiply(a: Vector, value: number, out?: Vector): Vector {
        if (!out) out = new Vector();
        out.x = a.x * value;
        out.y = a.y * value;
        return out;
    }

    static dot(a: Vector, b: Vector) {
        return a.dot(b);
    }

    static lerp(a: Vector, b: Vector, t: number, out?: Vector): Vector {
        if (!out) out = new Vector();
        out.x = a.x + t * (b.x - a.x);
        out.y = a.y + t * (b.y - a.y);
        return out;
    }

    static clone(source: Vector, out?: Vector): Vector {
        if (out) {
            out.copyFrom(source);
            return out;
        }

        return new Vector(source.x, source.y);
    }

    static fromPoint(pt: IPoint, out?: Vector) {
        if (!out) out = new Vector();
        out.x = pt.x;
        out.y = pt.y;
        return out;
    }

    static round(source: Vector, out?: Vector): Vector {
        if (!out) out = new Vector();
        out.x = Math.round(source.x);
        out.y = Math.round(source.y);
        return out;
    }

    static readonly zero = new Vector(0, 0);
    static readonly one = new Vector(1, 1);
    static readonly right = new Vector(1, 0);
    static readonly left = new Vector(-1, 0);
    static readonly up = new Vector(0, -1);
    static readonly down = new Vector(0, 1);
}

File: BattlrNet/Code/Game/Vector.cs
using System;

namespace BattlrNet.Code.Game
{
    public struct Vector
    {
        public float X;
        public float Y;

        public Vector(float x, float y)
        {
            X = x;
            Y = y;
        }

        public Vector Round()
        {
            return new Vector((float)Math.Round(X), (float)Math.Round(Y));
        }

        public float Length()
        {
            return (float)Math.Sqrt(X * X + Y * Y);
        }

        public float LengthSquared()
        {
            return X * X + Y * Y;
        }

        public static void Lerp(Vector a, Vector b, float t, ref Vector result)
        {
            result.X = a.X + t * (b.X - a.X);
            result.Y = a.Y + t * (b.Y - a.Y);
        }

        public static Vector Lerp(Vector a, Vector b, float t)
        {
            var v = new Vector();
            Lerp(a, b, t, ref v);
            return v;
        }

        public static Vector operator +(Vector a, Vector b)
        {
            return new Vector(a.X + b.X, a.Y + b.Y);
        }

        public static Vector operator -(Vector a, Vector b)
        {
            return new Vector(a.X - b.X, a.Y - b.Y);
        }

        public static Vector operator *(Vector a, float v)
        {
            return new Vector(a.X * v, a.Y * v);
        }

        public static Vector operator /(Vector a, float v)
        {
            return new Vector(a.X / v, a.Y / v);
        }

        public static readonly Vector Zero = new Vector(0, 0);
        public static readonly Vector One = new Vector(1, 1);
        public static readonly Vector Right = new Vector(1, 0);
        public static readonly Vector Left = new Vector(-1, 0);
        public static readonly Vector Up = new Vector(0, -1);
        public static readonly Vector Down = new Vector(0, 1);
    }
}
